### PR TITLE
Implement Return Type Notation (RTN)'s path form in where clauses

### PR DESCRIPTION
**Pull request by compiler-errors**
Implement return type notation (RTN) in path position for where clauses. We already had RTN in associated type position ([e.g.](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=627a4fb8e2cb334863fbd08ed3722c09)), but per [the RFC](https://rust-lang.github.io/rfcs/3654-return-type-notation.html#where-rtn-can-be-used-for-now):

> As a standalone type, RTN can only be used as the Self type of a where-clause [...]

Specifically, in order to enable code like:

```rust
trait Foo {
    fn bar() -> impl Sized;
}

fn is_send(_: impl Send) {}

fn test<T>()
where
    T: Foo,
    T::bar(..): Send,
{
    is_send(T::bar());
}
```

* In the resolver, when we see a `TyKind::Path` whose final segment is `GenericArgs::ParenthesizedElided` (i.e. `(..)`), resolve that path in the *value* namespace, since we're looking for a method.
* When lowering where clauses in HIR lowering, we first try to intercept an RTN self type via `lower_ty_maybe_return_type_notation`. If we find an RTN type, we lower it manually in a way that respects its higher-ranked-ness (see below) and resolves to the corresponding RPITIT. Anywhere else, we'll emit the same "return type notation not allowed in this position yet" error we do when writing RTN in every other position.
* In `resolve_bound_vars`, we add some special treatment for RTN types in where clauses. Specifically, we need to add new lifetime variables to our binders for the early- and late-bound vars we encounter on the method. This implements the higher-ranked desugaring [laid out in the RFC](https://rust-lang.github.io/rfcs/3654-return-type-notation.html#converting-to-higher-ranked-trait-bounds).

This PR also adds a bunch of tests, mostly negative ones (testing error messages).

In a follow-up PR, I'm going to mark RTN as no longer incomplete, since this PR basically finishes the impl surface that we should initially stabilize, and the RFC was accepted.

cc [RFC 3654](https://github.com/rust-lang/rfcs/pull/3654) and https://github.com/rust-lang/rust/issues/109417

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 1]
```diff
@@ -821,10 +821,11 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     ///
     /// `ty_param_def_id` is the `LocalDefId` of the type parameter.
     #[instrument(level = "debug", skip_all, ret)]
-    fn probe_single_ty_param_bound_for_assoc_ty(
+    fn probe_single_ty_param_bound_for_assoc_item(
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:824]: docs needs s/associated type/associated item/

**Comment by jackh726** [comment]
Sorry! Yes, I'll take a look sometime this week.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs [hunk 2]
```diff
@@ -525,6 +464,269 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
+                    });
+                    ty::Const::new_error(tcx, guar).into()
+                }
+            };
+            num_bound_vars += 1;
+            arg
+        });
+
+        // Next, we need to check that the return-type notation is being used on
+        // an RPITIT (return-position impl trait in trait) or AFIT (async fn in trait).
```
Comment (jackh726) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:708]: This is...not a restriction I expect given the RFC:
>The primary use case is to add bounds such as Send to the futures returned by async fns in traits and -> impl Future functions, but they work for any trait function defined with return-position impl trait (e.g., where T: Factory<widgets(..): DoubleEndedIterator> would also be valid).

At the very least, this should have a test.

**Review by jackh726 — COMMENTED** [review]
I think I probably want to do a second pass over this, but I do think some more tests here are needed.

File: compiler/rustc_resolve/src/late.rs [hunk 3]
```diff
@@ -781,7 +789,20 @@ impl<'ra: 'ast, 'ast, 'tcx> Visitor<'ast> for LateResolutionVisitor<'_, 'ast, 'r
             }
             TyKind::Path(qself, path) => {
                 self.diag_metadata.current_type_path = Some(ty);
-                self.smart_resolve_path(ty.id, qself, path, PathSource::Type);
+
+                // If we have a path that ends with `(..)`, then it must be
+                // return type notation. Resolve that path in the *value*
+                // namespace.
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_resolve/src/late.rs:795]: What specifically would you like tested here? When there's a conflicting definition in the type namespace?

File: tests/ui/suggestions/let-binding-init-expr-as-ty.stderr [hunk 4]
```diff
@@ -6,29 +6,22 @@ LL |     let foo: i32::from_be(num);
...
+error: return type notation not allowed in this position yet
   --> $DIR/let-binding-init-expr-as-ty.rs:2:14
    |
 LL |     let foo: i32::from_be(num);
    |              ^^^^^^^^^^^^^^^^^
-   |
-help: if there were a trait named `Example` with associated type `from_be` implemented for `i32`, you could use the fully-qualified path
-   |
-LL |     let foo: <i32 as Example>::from_be;
-   |              ~~~~~~~~~~~~~~~~~~~~~~~~~
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply tests/ui/suggestions/let-binding-init-expr-as-ty.stderr:29]: I don't consider this to be a diagnostic regression. This suggestion is just as nonsense as suggesting RTN; the correct suggestion would suggest to turn `foo:` into `foo =`, but that requires way too much lookahead.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs (hunk 2 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:708]: This is a restriction specifically called out in the RFC: https://rust-lang.github.io/rfcs/3654-return-type-notation.html#rtn-only-applies-to-afit-and-rpitit-methods

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:708]: I've extended `tests/ui/associated-type-bounds/return-type-notation/non-rpitit.rs` to exercise this codepath, which was already exercised for the associated-type-bound flavor of RTN.

File: compiler/rustc_resolve/src/late.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_resolve/src/late.rs:795]: I've added `tests/ui/associated-type-bounds/return-type-notation/namespace-conflict.rs` to ensure that we're not resolving to something in the type namespace (namely, a GAT).

IDK about `method()` by itself, but that needs to be resolved in the type namespace for bare `Fn()` trait objects. If you meant `T::method()` or `<T as Trait>::method()`: the diagnostics for the former (which we lower as a type-dependent path in HIR) are nice, but the latter (which we lower in resolve) are not. I can fix that as a follow-up, I guess :)

File: compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs [hunk 5]
```diff
@@ -1839,6 +1844,167 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
...
+                }
+                ty::GenericParamDefKind::Const { .. } => ty::BoundVariableKind::Const,
+            }
+        }));
+        bound_vars.extend(self.tcx.fn_sig(item_def_id).instantiate_identity().bound_vars());
+
+        // SUBTLE: Stash the old bound vars onto the *item segment* before appending
+        // the new bound vars. We do this because we need to know how many bound vars
+        // are present on the binder explicitly (i.e. not return-type-notation vars)
+        // to do bound var shifting correctly in HIR lowering.
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:2002]: This is exercised by `tests/ui/associated-type-bounds/return-type-notation/higher-ranked-bound-works.rs`.

File: compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs [hunk 6]
```diff
@@ -1839,6 +1844,167 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
...
+    // For example, given
+    // ```
+    // trait Foo {
+    //     async fn x<'r, T>();
+    // }
+    // ```
+    // and a bound that looks like:
+    //    `for<'a, 'b> <T as Trait<'a>>::x(): Other<'b>`
+    // this is going to expand to something like:
+    //    `for<'a, 'b, 'r, T> <T as Trait<'a>>::x::<'r, T>::{opaque#0}: Other<'b>`.
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs]: This comment is a bit out of date. We don't support type params anyways, but `tests/ui/associated-type-bounds/return-type-notation/higher-ranked-bound-works.rs` exercises when we have bound vars in both the trait ref, the method, and the goal.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs (hunk 2 above)
Comment (jackh726) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:708]: > This is a restriction specifically called out in the RFC: https://rust-lang.github.io/rfcs/3654-return-type-notation.html#rtn-only-applies-to-afit-and-rpitit-methods

Hmm, that quoted text is from the RFC, the summary section. That should probably be amended.

File: compiler/rustc_resolve/src/late.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jackh726) [reply compiler/rustc_resolve/src/late.rs:795]: Can you add a test for what diagnostic we emit if the user writes `method(): Send`?

**Review by jackh726 — APPROVED** [review]
Just the one comment remaining about adding a test for the mistyped `method(): Send` diagnostic, then r=me.

File: compiler/rustc_resolve/src/late.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_resolve/src/late.rs:795]: Should be covered by `tests/ui/associated-type-bounds/return-type-notation/bad-inputs-and-output.rs` and `tests/ui/associated-type-bounds/return-type-notation/not-a-method.rs`. The diagnostics could definitely use work on the resolver side, but I'll defer that to a follow-up.

File: tests/ui/associated-type-bounds/return-type-notation/not-a-method.stderr [hunk 7]
```diff
@@ -0,0 +1,49 @@
+error[E0575]: expected function, found function `function`
```
Comment (compiler-errors) [inline tests/ui/associated-type-bounds/return-type-notation/not-a-method.stderr:1]: Namely, this is a consequence of E0575 calling associated functions "functions". 

As a follow-up, I will do like https://github.com/rust-lang/rust/pull/130414 and probably split these resolver errors into their own error codes, so they can have tailored explanations.

**Pull request merged**